specs/distros.rst

```diff
@@ -0,0 +1,580 @@
+
+
+Users subnamespace
+------------------
+
+The base class for this namespace can look like this
+
+
+     class UserBase(ABCMeta):
+
+         def groups(self):
+             """Get all the user groups from the instance.
+
+             Similar with network.routes() et al, that is
+
+             >>> groups = distro.users.groups()
+             GroupContainer(Group(...), Group(....), ...)
+             # create a new group
+             >>> group = distro.users.Group.create(name)
+             # Add new members to a group
+             >>> group.add(member)
+             # Add a new group
+             >>> groups.add(group)
+             # Remove a group
+             >>> groups.delete(group)
+             # Iterate groups
+             >>> list(groups)
+
+             This gets rid of ``cloudinit.distros.Distro.create_group``,
+             which creates a group and adds member to it as well and it get rids of
+             ``cloudbaseinit.osutils.add_user_to_local``.
+             """
+
+       def users(self):
+             """Get all the users from the instance.
+
+             Using the same idion as for :meth:`routes` and :meth:`groups`.
+
+             >>> users = distro.users.users()
+             # containment (cloudbaseinit.osutils.user_exists)
+             >>> user in users
+             # Iteration
+             >>> for i in user: print(user)
+             # Add a new user
+             >>> user = users.create(username, password, password_expires=False)
+             """
+
+     class User:
+         """ Abstracts away user interaction.
+
+         # get the home dir of an user
+         >>> user.home()
+         # Get the password (?)
+         >>> user.password
+         # Set the password
+         >>> user.password = ....
+         # Get an instance of an User from a name
+         >>> user = distros.users.User.from_name('user')
+         # Disable login password
+         >>> user.disable_login_password()
+         # Get ssh keys
+         >>> keys = user.ssh_keys()
+
+         Posix specific implementations might provide some method
+         to operate with '/etc/sudoers' file.
+         """
+
+.. note::
+
+   TODO: what is cloudinit.distros.get_default_user?
+
+Packaging namespace
+-------------------
+
+This object is a thin layer over Distro specific packaging utilities,
+used in cloudinit through ``distro.Distro.package_command``.
+Instead of passing strings with arguments, as it currently does,
+we could have a more OO approach:
+
+      >>> distro.packaging.install(...)
+
+      # cloudinit provides a ``package_command`` and an ``update_package_sources`` method,
+      # which is:
+      #          self._runner.run("update-sources", self.package_command,
+      #                   ["update"], freq=PER_INSTANCE)
+      #  distro.packaging.update() can be a noop operation if it was already called
+      >>> distro.packaging.update(...)
+
+ On Windows side, this can be implemented with OneGet.
```

Comment:
Ok dokie; thx. I've used chocalatey, didn't know microsoft was behind a packaging/package manager effort though; the things u learn :-P